client/gutenberg/editor/components/post-publish-button-or-toggle/index.jsx

```diff
@@ -0,0 +1,217 @@
+		 * for a particular role (see https://codex.wordpress.org/Post_Status):
+		 *
+		 * - is published
+		 * - is scheduled to be published
+		 * - is pending and can't be published (but only for viewports >= medium).
+		 * 	 Originally, we considered showing a button for pending posts that couldn't be published
+		 * 	 (for example, for an author with the contributor role). Some languages can have
+		 * 	 long translations for "Submit for review", so given the lack of UI real estate available
+		 * 	 we decided to take into account the viewport in that case.
+		 *  	 See: https://github.com/WordPress/gutenberg/issues/10475
+		 *
+		 * 2) Then, in small viewports, we'll show a TOGGLE.
+		 *
+		 * 3) Finally, we'll use the publish sidebar status to decide:
+		 *
+		 * - if it is enabled, we show a TOGGLE
+		 * - if it is disabled, we show a BUTTON
+		 */
+		if (
+			isPublished ||
+			( isScheduled && isBeingScheduled ) ||
+			( isPending && ! hasPublishAction && ! isLessThanMediumViewport )
+		) {
+			component = IS_BUTTON;
+		} else if ( isLessThanMediumViewport ) {
+			component = IS_TOGGLE;
+		} else if ( isPublishSidebarEnabled ) {
+			component = IS_TOGGLE;
+		} else {
+			component = IS_BUTTON;
+		}
+
+		return (
+			<Fragment>
+				<PostPublishButton
+					forceIsDirty={ forceIsDirty }
+					forceIsSaving={ forceIsSaving }
+					isOpen={ isPublishSidebarOpened }
+					isToggle={ component === IS_TOGGLE }
+					onToggle={ togglePublishSidebar }
+				/>
+				{ showEmailVerificationDialog && (
+					<VerifyEmailDialog onClose={ this.closeVerifyEmailDialog } />
+				) }
+			</Fragment>
+		);
+	}
+}
+
+export default compose(
+	withSelect( select => ( {
+		hasPublishAction: get(
+			select( 'core/editor' ).getCurrentPost(),
+			[ '_links', 'wp:action-publish' ],
+			false
+		),
+		isBeingScheduled: select( 'core/editor' ).isEditedPostBeingScheduled(),
+		isPending: select( 'core/editor' ).isCurrentPostPending(),
+		isPublished: select( 'core/editor' ).isCurrentPostPublished(),
+		isPublishSidebarEnabled: select( 'core/editor' ).isPublishSidebarEnabled(),
+		isPublishSidebarOpened: select( 'core/edit-post' ).isPublishSidebarOpened(),
+		isScheduled: select( 'core/editor' ).isCurrentPostScheduled(),
+	} ) ),
+	withDispatch( dispatch => {
+		const { togglePublishSidebar } = dispatch( 'core/edit-post' );
+		const { createWarningNotice } = dispatch( 'core/notices' );
+		const { lockPostSaving, unlockPostSaving } = dispatch( 'core/editor' );
+		return {
+			togglePublishSidebar,
+			createWarningNotice,
+			lockPostSaving,
+			unlockPostSaving,
+		};
+	} ),
+	withViewportMatch( { isLessThanMediumViewport: '< medium' } ),
+	connect( state => {
+		const siteId = getSelectedSiteId( state );
+
+		// do not allow publish for unverified e-mails, but allow if the site is VIP, or if the site is unlaunched
+		const userNeedsVerification =
+			! isCurrentUserEmailVerified( state ) &&
+			! isVipSite( state, siteId ) &&
+			! isUnlaunchedSite( state, siteId );
```

Comment:
Testing the PR I've got into a funny case that somehow "breaks" these checks.

- Register new user.
- Keep the email unverified.
- Create a new post.
- Publish it: there are no unverified email warnings, as the site is "unlaunched".

This is correct: we allow unverified users to publish, as long as their site is not yet public.

Though, to launch a site, the user first needs to verify their email.

So, I wonder: when can it happen that a user has both a verified email and an unlaunched site—or viceversa?
Those two cases would be the only two that would enable the whole `userNeedsVerification` flow, with the warning and the dialog, etc. (in both Gutenberg and Classic)

cc @scruffian for adding `isUnlaunchedSite` to the equivalent of this in `EditorGroundControl` in #27962.

Reply:
It's possible for a user to have an unlaunched site and a verified email address. Since the process to launch a site would be to first verify the email address and then launch the site, any users who are part way through that process will be in that state.

The reverse state should be impossible - we should never allow users with unverified email addresses to launch a site.

Does that answer your question?

Reply:
Absolutely yes, thank you very much @scruffian!
I didn't proceed with the test ~because I'm that lazy~ to keep the user email unverified, and somehow just assumed verifying the email would also automatically launch (which would definitely be odd).

Reply:
(Sorry for the delay from my previous answer, I just now got to test it)

So, what happens then is that:
- Verified user can have launched and unlaunched sites.
- Unverified user can only have unlaunched sites.

In other words, this would never be true:
```es6
const userNeedsVerification =
	! isCurrentUserEmailVerified( state ) &&
	! isUnlaunchedSite( state, siteId );
```

In other words, we don't need the unverified email notice and the logic that prevents publishing a post, for both Guten and Classic.

Am I wrong on this?
Like, it's almost EOD and I've got the flu, I might very well have missed something. 😄

Reply:
> Unverified user can only have unlaunched sites.

I think this is not true. According to #27962, only private sites are unlaunched, so an unverified user can have a launched site if it is public.